site/content/blog/2020-09-22-sites-tree-modifiers/index.md

```diff
@@ -0,0 +1,125 @@
+
+In this case both URLs will end up in the same Sites Tree node, which means that in practice only one of them will be attacked. The other URL will not be attacked and so any vulnerabilities specific to that action will not be found.
+
+In ZAP the solution to this is to define the ‘action’ parameter as a [Structural Parameter](/docs/desktop/start/features/structparams/) ie a parameter whose value actually affects the structure of an app. Once you have done that (and revisited the URLs) then they will appear in 2 different nodes and ZAP will attack them separately.
+
+Conversely we have the case where a URL path element is actually part of the data and not part of the structure of the app.
+For example the following URLs could all represent the same functionality if the second path element (companyX) is actually data:
+
+* `https://www.example.com/app/company1/aaa?ddd=eee` 
+* `https://www.example.com/app/company2/aaa?ddd=fff` 
+* `https://www.example.com/app/company3/aaa?ddd=ggg`
+
+In this case all 3 URLs will end up in unique nodes and ZAP will attack each of them even though it doesn't really need to. This probably is not a big problem if there are just 3 instances, but in most cases this data will be coming from a database so there could be a huge number of such nodes.
+
+The solution to this is to define the relevant path element as [Data Driven Content](/docs/desktop/start/features/ddc/) ie path elements that really contain data rather than represent part of the app structure. Once you have done that (and revisited the URLs) then they will all be represented as just 1 node and ZAP will only attack them once.
+
+You will need to define both Structural Parameters and Data Driven Nodes manually - ZAP does not currently have the capability to automatically detect these situations.
+
+## Custom Data Formats
+
+We have yet another problem, and one that ZAP 2.9.0 cannot effectively solve.
+
+The above techniques only work if ZAP understands the structure of the data.
+While we do our best to make sure that ZAP can always understand the structure of the most common data formats there will always be custom formats that we never get to hear about or new formats that are introduced after a ZAP release.
+
+ZAP 2.9.0 does allow you to attack these custom formats as it provides Input Vector Scripts which allow anyone to write ZAP scripts to handle these formats. However these scripts cannot change how URLs are represented in the Sites tree which means that we end up with the same problems as before.
+
+# The Solution
+
+This will be solved in the next version of ZAP (2.10.0) which extends the definition of [Input Vector Scripts](https://github.com/zaproxy/community-scripts/tree/master/variant) to allow them to change the way URLs are represented in the sites tree. This means that anyone can write a script which will allow their app to be correctly represented in the Sites Tree and therefore be attacked effectively.
```

Comment:
```suggestion
This will be solved in the next version of ZAP (2.10.0) which extends the definition of [Input Vector Scripts](https://github.com/zaproxy/community-scripts/tree/master/variant) to allow them to change the way URLs are represented in the Sites Tree. This means that anyone can write a script which will allow their app to be correctly represented in the Sites Tree and therefore be attacked effectively.
```